### PR TITLE
Add support for parallelizing processing parquet files across workers and nodes.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

The main challenge with parquet files is to make sure each worker across all nodes processes the quantity of data. This PR introduces a lazy slice to make sure all workers process the exact same number of rows.

This PR introduces a ParquetReader to distribute processing parquet files across workers and nodes with ease.

```python
import os
from multiprocessing.pool import ThreadPool
from lightning.data import optimize
from lightning.data.processing.readers import ParquetReader
from lightning.data.processing.image import download_image_with_retry
from lightning.data.processing.utilities import SuppressStdoutStderr
from PIL import Image
from time import sleep

input_dir = "the-eye.eu/public/AI/cah/laion400m-met-release/laion400m-meta"
parquet_files = [os.path.join(input_dir, f) for f in os.listdir(input_dir) if f.endswith(".parquet")]

def process(row):
    image_id, url, text, height, width, image_license, nsfw, similarity = row
    img, err = download_image_with_retry(0, url, timeout=5)
    if err:
        return None, err

    try:
        return [image_id, img, text, image_license, nsfw, similarity], None
    except Exception:
        return None, err

class Fetcher:

    def __init__(self, max_threads=32):
        self.max_threads = max_threads
        self.stored = 0
        self.skipped = 0

    def __call__(self, df):
        print(self.skipped, self.stored)
        rows = [list(row) for row in df.iter_rows() if row[0] is not None]
        with ThreadPool(self.max_threads) as thread_pool:
            for row, err in thread_pool.imap_unordered(process, rows):
                if err is not None:
                    self.skipped += 1
                    continue

                if row[1] is not None:
                    try:
                        row[1] = Image.open(row[1]).resize((224, 224))
                    except:
                        self.skipped += 1
                        continue
                yield row
                self.stored += 1

optimize(
    fn=Fetcher(max_threads=16),
    inputs=parquet_files,
    output_dir="/teamspace/datasets/laion400m",
    num_workers=os.cpu_count(),
    reader=ParquetReader(num_rows=2048, to_pandas=False),
    chunk_bytes="64MB",
)
```

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda

### CODE REVIEW

File: src/lightning/data/processing/image.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import io
+from typing import Optional, Tuple
+
+from lightning_utilities.core.imports import RequirementCache
+
+_HTTPX_AVAILABLE = RequirementCache("httpx")
+
+# Credit to the https://github.com/rom1504/pytorch Github repo
```
Comment (lantiga) [inline src/lightning/data/processing/image.py:8]: link 404's
Also we need to mention the license here

File: src/lightning/data/processing/readers.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+_PYARROW_AVAILABLE = RequirementCache("pyarrow")
+
+
+class BaseReader(ABC):
+
+    def get_num_nodes(self) -> int:
+        return int(os.getenv("DATA_OPTIMIZER_NUM_NODES", 1))
+
+    @abstractmethod
+    def to_workers_user_items(self, items: List[Any], num_workers: int) -> List[List[Any]]:
```
Comment (lantiga) [inline src/lightning/data/processing/readers.py:21]: Not sure `user` is what we want for the name. Maybe `shard_items` or `items_to_workers`?

**Review by lantiga — APPROVED** [review]
LGTM, added two comments

File: src/lightning/data/processing/image.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/processing/image.py:8]: done.

File: src/lightning/data/processing/readers.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/processing/readers.py]: done !

**Review by dmitsf — APPROVED** [review]

**Pull request merged**